<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Исследование-данных-на-предмет-пропусков,-дубликатов,-явных-ошибок" data-toc-modified-id="Исследование-данных-на-предмет-пропусков,-дубликатов,-явных-ошибок-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Исследование данных на предмет пропусков, дубликатов, явных ошибок</a></span></li><li><span><a href="#Подготовка-категориальных-данных-методом-OHE" data-toc-modified-id="Подготовка-категориальных-данных-методом-OHE-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка категориальных данных методом OHE</a></span></li><li><span><a href="#Разделение-на-тестовую-и-учебную-выборки" data-toc-modified-id="Разделение-на-тестовую-и-учебную-выборки-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Разделение на тестовую и учебную выборки</a></span></li><li><span><a href="#Разделение-учебной-выборки-на-тренировочную-и-валидационную" data-toc-modified-id="Разделение-учебной-выборки-на-тренировочную-и-валидационную-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Разделение учебной выборки на тренировочную и валидационную</a></span></li><li><span><a href="#Стандартизация-признаков" data-toc-modified-id="Стандартизация-признаков-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Стандартизация признаков</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Применение-дерева-решений" data-toc-modified-id="Применение-дерева-решений-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Применение дерева решений</a></span></li><li><span><a href="#Применение-логистической-регрессии" data-toc-modified-id="Применение-логистической-регрессии-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Применение логистической регрессии</a></span></li><li><span><a href="#Применение-случайного-леса" data-toc-modified-id="Применение-случайного-леса-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Применение случайного леса</a></span></li><li><span><a href="#Поиск-оптимальных-параметров-случайного-леса" data-toc-modified-id="Поиск-оптимальных-параметров-случайного-леса-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Поиск оптимальных параметров случайного леса</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span></li><li><span><a href="#Попробуем-увеличение-выборки" data-toc-modified-id="Попробуем-увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Попробуем увеличение выборки</a></span></li><li><span><a href="#Попробуем-уменьшение-выборки" data-toc-modified-id="Попробуем-уменьшение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Попробуем уменьшение выборки</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Задача: построить модель  с предельно большим значением *F1*-меры (0.59). Проверьте *F1*-меру на тестовой выборке самостоятельно. Измерить *AUC-ROC* метрику. 

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

**План работ:**  
1. Подготовка данных:   <br>
    1.1. Исследование данных на предмет пропусков, дубликатов, явных ошибок.<br>
    1.2. Подготовка категориальных данных методом OHE<br>
    1.3. Стандартизация признаков.<br>
    1.4. Разделение данных на тестовые и учебные.<br>
    1.5. Разделение  учебных данных на тренировочные и валидационные. <br>
2. Исследование задачи<br>
    2.1. Применение дерева решений<br>
    2.2. Применение логистической регрессии<br>
    2.3. Применение случайного леса.<br>
    2.4. Поиск наилучших параметров случайного леса.  <br>
3. Борьба с дисбалансом<br>
    3.1. Взвешивание классов<br>
    3.2. Увеличение выборки<br>
    3.3. Уменьшение выборки<br>
4. Тестиование оптимальной модели<br>
5. Выводы
        

## Подготовка данных

### Исследование данных на предмет пропусков, дубликатов, явных ошибок

In [1]:
#Импорт библиотек для дальнейшей работы
import pandas as pd
#import pandas_profiling
import numpy as np

pd.options.display.float_format = '{:.2f}'.format 
#в дальнейшей таблице очень много знаков после запятой, ограраничим представление двумя.

pd.set_option('display.max_columns', 50) 
pd.set_option('display.max_rows', 500) 

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:90% !important; }</style>"))


import matplotlib.pyplot as plt 
import numpy as np
from scipy import stats as st 
from matplotlib import pyplot
import seaborn as sns
from matplotlib import cm
import random

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score
from sklearn.utils import shuffle
from sklearn.metrics import roc_auc_score 

C:\Users\lenovo\AppData\Local\Temp\ipykernel_19200\463960154.py:12: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [19]:
try:
    data = pd.read_csv('/datasets/Churn.csv')
except:
    data = pd.read_csv('Churn.csv')
    
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0    2.00      0.00              1          1               1   
1    1.00  83807.86              1          0               1   
2    8.00 159660.80              3          1               0   
3    1.00      0.00              2          0               0   
4    2.00 125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [20]:
data.shape

(10000, 14)

У нас на рассмотрении таблица в 10 000 строк. 

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Среди всех данных категорийными, требующими перевода в численными являются Geography, Gender. Также из обработки можно в будущем удалить Surname (фамилия человека на его кредитоспособность никак не влияет). 

In [22]:
data.isna().mean()

RowNumber         0.00
CustomerId        0.00
Surname           0.00
CreditScore       0.00
Geography         0.00
Gender            0.00
Age               0.00
Tenure            0.09
Balance           0.00
NumOfProducts     0.00
HasCrCard         0.00
IsActiveMember    0.00
EstimatedSalary   0.00
Exited            0.00
dtype: float64

In [23]:
data['Tenure'].unique()

array([ 2.,  1.,  8.,  7.,  4.,  6.,  3., 10.,  5.,  9.,  0., nan])

In [24]:
data.query('Tenure.isna()')

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN      0.00              3          1               0   
48       NaN 103391.38              1          0               1   
51       NaN 146050.97              2          0               0   
53       NaN 125561.97              1          0               0   
60       NaN 136857.00              1          0               0   
...      ...       ...            ...        ...             ...   
9944     NaN 190409.34              2          1               1   
9956     NaN  85216.61              1          1               0   
9964     NaN 117593.48              2          0               0   
9985     NaN 123841.49              2          1               0   
9999     NaN 130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

In [25]:
data.query('Tenure.notna()')

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9994       9995    15719294       Wood          800    France  Female   29   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   

      Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2.00      0.00              1          1               1   
1       1.00  83807.86              1          0               1   
2       8.00 159660.80              3          1               0   
3       1.00      0.00              2          0               0   
4       2.00 125510.82              1          1               1   
...      ...       ...            ...        ...             ...   
9994    2.00      0.00              2          0               0   
9995    5.00      0.00              2          1               0   
9996   10.00  57369.61              1          1               1   
9997    7.00      0.00              1          0               1   
9998    3.00  75075.31              2          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9994        167773.55       0  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  

[9091 rows x 14 columns]

9% строк в Tenure (время пользования банком) являются  незаполненными. Поскольку это все же численные данные, заменим Nan на заглушку.

In [30]:
data['Tenure'] = data['Tenure'].fillna(value = 999)

In [31]:
data['Tenure'].unique()

array(['2.0', '1.0', '8.0', '7.0', '4.0', '6.0', '3.0', '10.0', '5.0',
       '9.0', '0.0', '999.0'], dtype=object)

In [32]:
data['Tenure'] = data['Tenure'].astype('str')

In [33]:
data.duplicated().sum()

0

Дублирующихся строк не обнаружено

In [34]:
data.describe()

RowNumber  CustomerId  CreditScore      Age   Balance  NumOfProducts  \
count   10000.00    10000.00     10000.00 10000.00  10000.00       10000.00   
mean     5000.50 15690940.57       650.53    38.92  76485.89           1.53   
std      2886.90    71936.19        96.65    10.49  62397.41           0.58   
min         1.00 15565701.00       350.00    18.00      0.00           1.00   
25%      2500.75 15628528.25       584.00    32.00      0.00           1.00   
50%      5000.50 15690738.00       652.00    37.00  97198.54           1.00   
75%      7500.25 15753233.75       718.00    44.00 127644.24           2.00   
max     10000.00 15815690.00       850.00    92.00 250898.09           4.00   

       HasCrCard  IsActiveMember  EstimatedSalary   Exited  
count   10000.00        10000.00         10000.00 10000.00  
mean        0.71            0.52        100090.24     0.20  
std         0.46            0.50         57510.49     0.40  
min         0.00            0.00            11.58     0.00  
25%         0.00            0.00         51002.11     0.00  
50%         1.00            1.00        100193.91     0.00  
75%         1.00            1.00        149388.25     0.00  
max         1.00            1.00        199992.48     1.00

Данные выглядят правдоподобно (возраст от 18 до 92, пользование банком от 0 до 10 лет, оклад от 50 до 200 тыс. руб.)

In [35]:
#Удалим столбец с фамилией
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis = 1)
data.head()

CreditScore Geography  Gender  Age Tenure   Balance  NumOfProducts  \
0          619    France  Female   42    2.0      0.00              1   
1          608     Spain  Female   41    1.0  83807.86              1   
2          502    France  Female   42    8.0 159660.80              3   
3          699    France  Female   39    1.0      0.00              2   
4          850     Spain  Female   43    2.0 125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [36]:
#Удалим столбец с фамилией
data = data.drop(['Surname', axis = 1)

SyntaxError: invalid syntax (3533631814.py, line 2)

Был удален столбец с фамилиями пользователей и заменены данные в Tenure (сроке пользвания услугами) на медианные.

### Подготовка категориальных данных методом OHE

In [37]:
#Преобразуем Geography и Gender прямым кодированием.
display(data['Geography'].unique())
display(data['Gender'].unique())


data_ohe = pd.get_dummies(data, drop_first=True)
data_ohe.head()
data_ohe.columns

array(['France', 'Spain', 'Germany'], dtype=object)

array(['Female', 'Male'], dtype=object)

Index(['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited', 'Geography_Germany',
       'Geography_Spain', 'Gender_Male', 'Tenure_1.0', 'Tenure_10.0',
       'Tenure_2.0', 'Tenure_3.0', 'Tenure_4.0', 'Tenure_5.0', 'Tenure_6.0',
       'Tenure_7.0', 'Tenure_8.0', 'Tenure_9.0', 'Tenure_999.0'],
      dtype='object')

В получившейся таблице если география не Германия и не Испания, то Франция, если не мужчина, то женщина.

In [38]:
#Разделим данные на target/features

target = data_ohe['Exited']
features = data_ohe.drop('Exited', axis=1)

### Разделение на тестовую и учебную выборки

In [39]:
#Выделим 25% тестовой выборки:
features_test, features_learn, target_test, target_learn = train_test_split(features, target, test_size=0.75, random_state=12345)

features_test.shape

(2500, 21)

### Разделение учебной выборки на тренировочную и валидационную

In [40]:
#Выделим 25% валидационной выборки:
features_train, features_valid, target_train, target_valid = train_test_split(features_learn, target_learn, test_size=0.25, random_state=12345)

print(features_train.shape)
print(features_valid.shape)
print(target_test.unique())

(5625, 21)
(1875, 21)
[0 1]


### Стандартизация признаков

In [41]:
#Выделим столбцы, которые нужно стандартизировать(численные)
features_train.columns

Index(['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Geography_Germany',
       'Geography_Spain', 'Gender_Male', 'Tenure_1.0', 'Tenure_10.0',
       'Tenure_2.0', 'Tenure_3.0', 'Tenure_4.0', 'Tenure_5.0', 'Tenure_6.0',
       'Tenure_7.0', 'Tenure_8.0', 'Tenure_9.0', 'Tenure_999.0'],
      dtype='object')

In [42]:
#Приведем все признаки к единму масштабу

#перечислим численные столбцы:
numeric = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'EstimatedSalary']

scaler = StandardScaler()
scaler.fit(features_train[numeric])

features_train[numeric] = scaler.transform(features_train[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])


In [43]:
#проверим, сработала ли стандартизация
display(features_test[numeric].head())
display(features_valid[numeric].head())
display(features_train[numeric].head())


CreditScore   Age  Balance  NumOfProducts  EstimatedSalary
7072         0.38 -0.17     0.94          -0.92            -1.29
5028        -1.77 -0.07     1.53          -0.92            -0.87
4509        -0.80 -0.36     1.02          -0.92             1.47
5151         1.29 -0.36     0.73          -0.92            -1.34
1443        -0.97  0.41    -1.22           2.55             0.41

CreditScore   Age  Balance  NumOfProducts  EstimatedSalary
5361        -0.63  0.70    -1.22           0.81             1.21
8838        -0.95 -0.55     0.52          -0.92             0.13
1612         0.16  1.28    -1.22          -0.92             0.26
8947         0.53  0.41     1.35          -0.92            -1.12
1989         0.47 -1.03     0.32          -0.92            -0.76

CreditScore   Age  Balance  NumOfProducts  EstimatedSalary
8026        -0.44 -0.55     0.24          -0.92             0.51
5808         0.36  2.05     0.65          -0.92             0.33
7161         0.42 -0.26     0.55          -0.92            -0.35
7133         0.97 -0.84     0.82          -0.92             0.65
4661         1.35  0.41    -1.22          -0.92            -1.50

## Исследование задачи

### Применение дерева решений

In [44]:
#Протестируем дерево решений на нашей выборке:
model = DecisionTreeClassifier(random_state=12345)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)


probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)


In [45]:
print(f'F1 метрика для валидационной выборки дерева решений: {f1_score(target_valid, predicted_valid)}')
print(f'AUC-ROC метрика для валидационной выборки дерева решений {auc_roc_valid}')

F1 метрика для валидационной выборки дерева решений: 0.5039164490861618
AUC-ROC метрика для валидационной выборки дерева решений 0.694212335189406


### Применение логистической регрессии

In [46]:
#Для контроля проверим логистической регрессией:
model = LogisticRegression(random_state=12345, solver='liblinear', class_weight = 'balanced')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)


In [47]:
print(f'F1 метрика для валидационной выборки логистической регрессии: {f1_score(target_valid, predicted_valid)}')
print(f'AUC-ROC метрика для валидационной выборки логистической регрессии {auc_roc_valid}')

F1 метрика для валидационной выборки логистической регрессии: 0.48725212464589235
AUC-ROC метрика для валидационной выборки логистической регрессии 0.7729959464527856


### Применение случайного леса

In [48]:
#Протестируем случайный лес для нашей выборки и посмотрим на тестовые значения
model = RandomForestClassifier(random_state=12345)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)



In [49]:
print(f'F1 метрика для валидационной выборки случайным лесом: {f1_score(target_valid, predicted_valid)}')
print(f'AUC-ROC метрика для валидационной выборки логистической регрессии {auc_roc_valid}')

F1 метрика для валидационной выборки случайным лесом: 0.5675213675213674
AUC-ROC метрика для валидационной выборки логистической регрессии 0.8387578979256225


### Поиск оптимальных параметров случайного леса

In [50]:
%%time


#Случайный лес подбор глубины
best_result = 0
best_depth = 0
for depth in range(1, 150, 1):
    model = RandomForestClassifier(n_estimators=10, max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    predict = model.predict(features_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_depth = depth

print(f'Лучшее значение f1: {best_result} для глубины {best_depth}')
    


Лучшее значение f1: 0.5502645502645502 для глубины 10
CPU times: total: 7.98 s
Wall time: 7.98 s


In [51]:
%%time 

best_result = 0
best_estimators = 5
for est in range(5, 200, 10):
    model = RandomForestClassifier(n_estimators=est, max_depth=10, random_state=12345)
    model.fit(features_train, target_train)
    predict = model.predict(features_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_estimators = est

print(f'Лучшее значение f1: {best_result} для количества деревьев (est) {est}')

Лучшее значение f1: 0.5367647058823529 для количества деревьев (est) 195
CPU times: total: 8.2 s
Wall time: 8.24 s


Посмотрим на тестовую выборку  с лучшими параметрами случайного леса:

In [52]:
model = RandomForestClassifier(max_depth = 10, n_estimators = 195, random_state=12345)
model.fit(features_train, target_train)
predicted_test = model.predict(features_test)
probabilities_test = model.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc_test = roc_auc_score(target_test, probabilities_one_test)


print(f'F1 метрика для тестовой выборки случайным лесом: {f1_score(target_test, predicted_test)}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом {auc_roc_test}')

F1 метрика для тестовой выборки случайным лесом: 0.5650557620817844
AUC-ROC метрика для  валидационной выборки случайным лесом 0.860868331228197


На тестовой выборке значение  F1 0,565 для случайного леса, на валидационной: 0,53.
Поработаем с дисбалансом классов, но теперь будем рассматривать результаты на тестовой и на валидационной выборке, поскольку мы нашли оптимальную модель. 

## Борьба с дисбалансом

### Взвешивание классов

In [53]:
model = RandomForestClassifier(max_depth = 10, n_estimators = 195, random_state=12345, class_weight = 'balanced')
model.fit(features_train, target_train)
predicted_test = model.predict(features_test)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)

print(f'F1 метрика для тестовой выборки случайным лесом со взвешенными классами: {f1_score(target_valid, predicted_valid)}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами {auc_roc_valid}')


F1 метрика для тестовой выборки случайным лесом со взвешенными классами: 0.5675213675213674
AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами 0.8536539626669744


In [54]:
%%time

best_result = 0
best_depth = 0
for depth in range(1, 150, 1):
    model = RandomForestClassifier(n_estimators=10, max_depth=depth, random_state=12345, class_weight = 'balanced')
    model.fit(features_train, target_train)
    predict = model.predict(features_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_depth = depth

best_result = 0
best_estimators = 5
best_auc_roc = 0
for est in range(5, 200, 10):
    model = RandomForestClassifier(n_estimators=est, max_depth=best_depth, random_state=12345, class_weight = 'balanced')
    model.fit(features_train, target_train)
    predict = model.predict(features_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_estimators = est
        best_auc_roc = auc_roc_valid
        
print(f'Лучшая модель случайного леса со взвешенными классами: est: {best_estimators}, max_depth: {best_depth}')
print(f'F1 метрика для валидационной выборки случайным лесом со взвешенными классами: {best_result}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами {best_auc_roc}')


Лучшая модель случайного леса со взвешенными классами: est: 105, max_depth: 9
F1 метрика для валидационной выборки случайным лесом со взвешенными классами: 0.6281407035175879
AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами 0.8531598915207294
CPU times: total: 16.2 s
Wall time: 16.2 s


Метрика на валидационной выборке равна 0,62 для взвешенных классов со случкйнам лесом: est:105, max_depth: 9. 
Это удовлетворительный результат, но поработаем с классами пристальнее.

### Попробуем увеличение выборки

In [55]:
#Используем функцию для увеличения выборки

def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled


In [56]:
#Подберем лучшее увеличение выборки для показателей нашей модели (est = 195, max_depth = 10):

best_result = 0
best_result_test = 0
best_upsample_rate = 1
best_valid_auc_roc = 0
for sample_rate in range(1, 20, 1):
    features_upsampled, target_upsampled = upsample(features_train, target_train, sample_rate)
    model = RandomForestClassifier(max_depth = 11, n_estimators = 195, random_state=12345)
    model.fit(features_upsampled, target_upsampled)
    predict = model.predict(features_test)
    predict_valid = model.predict(features_valid)
    f1 = f1_score(target_valid, predict_valid)
    f1_t = f1_score(target_test, predict)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)   
    if best_result < f1:
        best_result = f1
        best_result_test = f1_t
        best_upsample_rate = sample_rate
        best_valid_auc_roc = auc_roc_valid
print(f'F1 метрика для тестовой выборки случайным лесом: {best_result_test} с увеличением {best_upsample_rate} на тестовой выборке')
print(f'F1 метрика для валидационной выборки случайным лесом: {best_result} с увеличением {best_upsample_rate}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами {best_valid_auc_roc}')


F1 метрика для тестовой выборки случайным лесом: 0.6210350584307179 с увеличением 5 на тестовой выборке
F1 метрика для валидационной выборки случайным лесом: 0.618096357226792 с увеличением 5
AUC-ROC метрика для  валидационной выборки случайным лесом со взвешенными классами 0.8502134964369175


In [57]:
#Подберем модель случайного леса с upsample_rate = 5

%%time

best_result = 0
best_depth = 0
features_upsampled, target_upsampled = upsample(features_train, target_train, 5)
for depth in range(1, 150, 1):
    model = RandomForestClassifier(n_estimators=10, max_depth=depth, random_state=12345)
    model.fit(features_upsampled, target_upsampled)
    predict = model.predict(features_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_depth = depth

best_result = 0
best_estimators = 5
best_auc_roc = 0
for est in range(5, 200, 10):
    model = RandomForestClassifier(n_estimators=est, max_depth=best_depth, random_state=12345)
    model.fit(features_upsampled, target_upsampled)
    predict = model.predict(features_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc_valid = roc_auc_score(target_valid, probabilities_one_valid)
    f1 = f1_score(target_valid, predict)
    if best_result < f1:
        best_result = f1
        best_estimators = est
        best_auc_roc = auc_roc_valid
        
print(f'Лучшая модель случайного леса с увеличенной выборкой: est: {best_estimators}, max_depth: {best_depth}')
print(f'F1 метрика для валидационной  выборки случайным лесом с увеличенной выборкой: {best_result}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом с увеличенной выборкой: {best_auc_roc}')



UsageError: Line magic function `%%time` not found.


Мы добились увеличением классов в 4 раза параметра  тестовой выборки F1 = 0,637 (на валидационной F1: 0,61). До сих пор это наилучший результат. 

### Попробуем уменьшение выборки

In [ ]:
### Попробуем уменьшить выборку

def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat([features_zeros.sample(frac=0.1, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat([target_zeros.sample(frac=0.1, random_state=12345)] + [target_ones])
    features_downsampled  = shuffle(features_downsampled, random_state = 12345)
    target_downsampled = shuffle(target_downsampled, random_state = 12345)
    return features_downsampled, target_downsampled

features_downsampled, target_downsampled = downsample(features_train, target_train, 0.1)

model = RandomForestClassifier(max_depth = 11, n_estimators = 195, random_state=12345)
model.fit(features_downsampled, target_downsampled)
predicted_test = model.predict(features_test)
print(f'F1 метрика для тестовой выборки случайным лесом: {f1_score(target_test, predicted_test)}')

Уменьшение выборки ухудшает модель. Принимаем за лучшее из возможных способов обучение модели на основе случайного леса (195 est, max_depth = 12, увеличение выборки в 4 раза). Лучший параметр F1 = 0.637

## Тестирование модели

In [ ]:
#Еще раз применим модель с наилучшими параметрами и вычислим F1 на тестовой выборке
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

model = RandomForestClassifier(max_depth = 11, n_estimators = 195, random_state=12345)
model.fit(features_upsampled, target_upsampled)
predicted_test = model.predict(features_test)
probabilities_test = model.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc_test = roc_auc_score(target_test, probabilities_one_test)




print(f'F1 метрика для тестовой выборки случайным лесом: {f1_score(target_test, predicted_test)}')
print(f'AUC-ROC метрика для  валидационной выборки случайным лесом {auc_roc_test}')


Вывод: в наилучшей из полученных моделей среднее гармоническое между точностью и полнотой равно 0,637 (в абсолютно идеальной модели оно равно 1). Также эта модель имеет наивысшее значение AUC-ROC, что говорит о лучшем качестве модели (модель тем лучше, чем ближе параметр к 1). 

Любопытно отметить, что на тестовой выборке модель отработала лучше, чем на валидационной. Это позволяет сказать, что модель не переобучена и ей можно добавить данных в тренировочную выборку. 

## Выводы

Задачей ставилось создание обучающей модели, которая сможет предсказать, перестанет клиент пользоваться услугами банка, или нет.
Исходные данные включали в себя 10 000 позиций для клиентов от 18 до 92 лет, разное время пользующихся услугами банка и ушедшими (или не ушедшими. Требовалось добиться от модели метрики F1 (среднее гармоническое по точности и полноте) не менее 59%. <br>

1. Подготовка данных:   
    1.1. Промущенные значения в столбце Tenure (время пользования банком) заменены на медианные. Удалены фамилии пользователей как незначимые для обучения. <br>
    1.2. Категориальные данные подготовлены методом OHE.<br>
    1.3. Численные признаки стандартизированы. <br>
    1.4. Данные разделены на учебные 75% и тестовые 25%. <br>
    1.5. Учебные данные разделены на тренировочные 75% и валидационные 25%. <br>
2. Исследование задачи на валидационной выборке:<br>
    2.1. Исследовано дерево решений: F1: 0.49, AUC-ROC: 0.69 <br>
    2.2. Применение логистической регрессии: F1: 0.49, AUC-ROC: 0.77 <br>
    2.3. Применение случайного леса: F1: 0.56, AUC-ROC: 0.83<br>
    2.4. Найдены лучие параметры случайного леса (max_depth = 11, n_estimators = 195). <br>
    Результат на тестовой выборке лучшей модели случайного леса:  F1: 0.59, AUC-ROC: 0.86<br>
3. Борьба с дисбалансом<br>
    3.1. Взвешивание классов на валидационной выборке дало F1: 0.58, AUC-ROC: 0.85 <br>
    3.2. Увеличение выборки в 4 раза дало для валидационной выборки F1: 0.61, AUC-ROC: 0.85   <br>
    3.3. Уменьшение выборки к улучшению результатов не привело. <br>
4. Тестиование оптимальной модели: F1: 0.637, AUC-ROC: 0.86 <br>

Получена оптимальная удовлетворительная модель для ответа на вопрос о пользовании услугами банка на основании данных клиента. F1: 0.637, AUC-ROC: 0.86. Случайный лес с увеличением выборки в 4 раза,  max_depth = 11, n_estimators = 195.

Рекомендации:
Поскольку показатели на тестовой выборке выше, чем на валидационной, модель не является переобученной. Можно добавить данных для обучения. 
